SnP_Emulator.py

```diff
@@ -0,0 +1,273 @@
+                if self._state is 'hard_puff':
+                    ramp_threshold = self._puff_ramp_down
+                else:
+                    ramp_threshold = self._puff_ramp_up
+                if in_state > ramp_threshold:
+                    self._state = 'soft_puff'
+                    #self._ramp_wait = False
+        elif pressure <= self._ambient_pressure - self._deadband:
+            if self._ramp_wait == False:
+                self._ramp_wait = True
+                self._ramp_time = time.time()
+            else:
+                curr_time = time.time()
+                in_state = curr_time - self._ramp_time
+                if self._state is 'hard_sip':
+                    ramp_threshold = self._sip_ramp_down
+                else:
+                    ramp_threshold = self._sip_ramp_up
+                if in_state > ramp_threshold:
+                    self._state = 'soft_sip'
+                    #self._ramp_wait = False
+            # self._state = 'soft_sip'
+        else:
+            self._state = 'deadband'
+            self._ramp_wait = False
+
+    def run(self):
+        # method run when ::meth::'start()' is called (comes from base class)
+        while True:
+            s = [0]
+            read_serial = self._port.readline()
+            try:
+                s[0] = int(read_serial)
+                # while not data_queue.empty():
+                #     data_queue.get_nowait()
+                # data_queue.put(s[0])
+                # s[0] = time.time()
+                self._setPressure(s[0])
+                # print(s[0])
+                # want this sleep to be less than the period of the serial info (50 ms as of 1/20/20)
+                # so that the read actually blocks until there's new data
+                time.sleep(0.025)
+                # self._port.reset_input_buffer()
+                # maybe put timeout on serail port and catch timeout exception and continue
+            except ValueError:
+                pass
+    def _setPressure(self, pressure):
+        with self._press_lock:
+            self._curr_pressure = pressure
+        if self._params_set:
+            self._setState(pressure)
+        # print(pressure)
+    def getPressure(self):
+        with self._press_lock:
+            return self._curr_pressure
+    
+    
+    
+    def getState(self):
+        return self._state            
+
+    def setup(self):
+        # snp_state = SnPState(0, 0, 0, 0)
+        pressures = {'hard_sip': [],
+                    'soft_sip': [],
+                    'hard_puff': [],
+                    'soft_puff': []
+                    }
+        avg_ambient_pressure = 0
+        i = 0
+        while i < NUM_SAMPLES:
+            pressure = self.getPressure()
+            avg_ambient_pressure += pressure
+            i = i + 1
+            time.sleep(0.1)
+        avg_ambient_pressure = avg_ambient_pressure / NUM_SAMPLES
+        self._ambient_pressure = avg_ambient_pressure
```

Comment:
I think we will want to also add the option for the OT to input all of the values manually - this way if a profile somehow gets deleted a patient doesn't have to go through the entire setup procedure again.